app/uk/gov/hmrc/pillar2/controllers/ReadSubscriptionCacheController.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.pillar2.controllers
+
+import play.api.Logging
+import play.api.mvc.{Action, AnyContent, ControllerComponents}
+import uk.gov.hmrc.pillar2.controllers.auth.AuthAction
+import uk.gov.hmrc.pillar2.repositories.ReadSubscriptionCacheRepository
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ReadSubscriptionCacheController @Inject() (
+  repository:                ReadSubscriptionCacheRepository,
+  authenticate:              AuthAction,
+  cc:                        ControllerComponents
+)(implicit executionContext: ExecutionContext)
+    extends BackendController(cc)
+    with Logging {
+
+  def save(id: String): Action[AnyContent] = authenticate.async { implicit request =>
+    request.body.asJson.map { jsValue =>
+      repository.upsert(id, jsValue).map(_ => Ok)
+    } getOrElse Future.successful(EntityTooLarge)
+  }
+
+  def get(id: String): Action[AnyContent] = authenticate.async { implicit request =>
+    repository.get(id).map { response =>
+      logger.debug(
+        s"controllers.ReadSubscriptionCacheController.get: Response for request Id $id is $response"
+      )
+      response.map(Ok(_)).getOrElse(NotFound)
+    }
+  }
+
+  def remove(id: String): Action[AnyContent] = authenticate.async { implicit request =>
+    repository.remove(id).map { response =>
+      if (response) Ok else InternalServerError
```

Comment:
Add a unit test for when response is false.

Reply:
Done